openedx_learning/core/publishing/models.py

```diff
@@ -13,36 +24,258 @@
 
 
 class LearningPackage(models.Model):
+    """
+    Top level container for a grouping of authored content.
+
+    Each PublishableEntity belongs to exactly one LearningPackage.
+    """
+
     uuid = immutable_uuid_field()
     identifier = identifier_field()
     title = models.CharField(max_length=1000, null=False, blank=False)
-
     created = manual_date_time_field()
     updated = manual_date_time_field()
 
     def __str__(self):
-        return f"{self.identifier}: {self.title}"
+        return f"{self.identifier}"
 
     class Meta:
         constraints = [
             # LearningPackage identifiers must be globally unique. This is
             # something that might be relaxed in the future if this system were
             # to be extensible to something like multi-tenancy, in which case
             # we'd tie it to something like a Site or Org.
-            models.UniqueConstraint(fields=["identifier"], name="lp_uniq_identifier")
+            models.UniqueConstraint(
+                fields=["identifier"],
+                name="publishable_lp_uniq_identifier",
+            )
         ]
         verbose_name = "Learning Package"
         verbose_name_plural = "Learning Packages"
 
 
-class PublishLogEntry(models.Model):
+class PublishableEntity(models.Model):
+    """
+    This represents any publishable thing that has ever existed in a
+    LearningPackage. It serves as a stable model that will not go away even if
+    these things are later unpublished or deleted.
+
+    A PublishableEntity belongs to exactly one LearningPackage.
+
+    How to use this model
+    ---------------------
+
+    The publishing app understands that publishable entities exist, along with
+    their drafts and published versions. It has some basic metadata, such as
+    identifiers, who created it, and when it was created. It's meant to
+    encapsulate the draft and publishing related aspects of your content, but
+    the ``publishing`` app doesn't know anything about the actual content being
+    referenced.
+
+    You have to provide actual meaning to PublishableEntity by creating your own
+    models that will represent your particular content and associating them to
+    PublishableEntity via a OneToOneField with primary_key=True. The easiest way
+    to do this is to have your model inherit from PublishableEntityMixin.
+
+    Identifiers
+    -----------
+    The UUID is globally unique and should be treated as immutable.
+
+    The identifier field *is* mutable, but changing it will affect all
+    PublishedEntityVersions. They are locally unique within the LearningPackage.
+
+    If you are referencing this model from within the same process, use a
+    foreign key to the id. If you are referencing this PublishedEntity from an
```

Comment:
`id` is different from `identifier` right? Seems like potential for confusion. Could we at least list it below as an explicit model field instead of implicit?

Reply:
How set are we on the term "identifier" @ormsbee @bradenmacdonald  ? Given that we have two other fields with ID in the name (`id` and `uuid`), and given that we recommend that folks *don't* rely on the field:
> The identifier is the part that is
    most likely to be human-readable, and may be exported/copied, but try not to
    rely on it, since this value may change.

it feels less like an `identifier` to me, and more like `slug`, `anchor`, or `key`.

Reply:
I was borrowing the `identifier` terminology from QTI. I think I need to qualify that recommendation more. I think it's more like saying "the decision to rely on the `identifier` field is something that can be made at a higher level, but our apps don't assume it's stable"? So for us, UsageKeys (or something we derive UsageKeys from, since I don't want to hard-code the course key into it) would go here. Whether its human readable depends on the higher level system–some of our UsageKeys are, most aren't. But we want to make foreign keys in such a way that changing a UsageKey/CourseKey won't break all the things like it does today.

I think that `slug` is very specifically meant to be human-readable, and doesn't quite strike the right connotation here. Maybe `key`? I haven't really heard of `anchor` being used for this kind of thing outside of HTML.

Reply:
I like `key`

Reply:
Changed it to `key`.